streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java

```diff
@@ -26,12 +26,15 @@
 import org.apache.kafka.streams.processor.Punctuator;
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 
 public class ProcessorNode<K, V> {
 
     private final List<ProcessorNode<?, ?>> children;
+    private final Map<String, ProcessorNode<?, ?>> childByName;
```

Comment:
Do we still need `children` above, could we use `childByName.values` to replace it? Generally speaking having two structures that always need to be updated means unnecessary maintenance cost./

Reply:
I add a TODO. Atm, we need to keep both, because we still support forward by index -- for this, we need a list to maintain the correct index -- using only the `Map` would break backward compatibility.

Reply:
I see. Sounds great.